amethyst_animation/src/systems/sampling.rs

```diff
@@ -0,0 +1,198 @@
+            do_sampling(&sampler, &duration, transform);
+        }
+        Done => do_end_control(&control, transform),
+        _ => {}
+    }
+
+    // Update state for next iteration
+    control.state = new_state;
+}
+
+/// Called on samplers that have finished.
+fn do_end_control(control: &SamplerControl, transform: &mut LocalTransform) {
+    match control.end {
+        EndControl::Normal => match control.after {
+            RestState::Translation(tr) => transform.translation = tr.into(),
+            RestState::Rotation(r) => transform.rotation = r.into(),
+            RestState::Scale(s) => transform.scale = s.into(),
+        },
+        // looping is handled during duration update
+        _ => {}
+    }
+}
+
+/// Update durations, check if the sampler is finished, start new samplers, and check for aborted
+/// samplers.
+///
+/// ## Parameters
+///
+/// - `control`: sampler control object
+/// - `sampler`: sampler reference from control
+/// - `now`: synchronized `Instant` for the current frame
+///
+/// ## Returns
+///
+/// Will return the new state of the sampling, and optionally a new end control state (for looping)
+fn update_duration_and_check(
+    control: &SamplerControl,
+    sampler: &Sampler,
+    now: &Instant,
+) -> (ControlState, Option<EndControl>) {
+    use resources::ControlState::*;
+    // Update state with new duration
+    // Check duration for end of sampling
+    match control.state {
+        // requested sampling => start interpolating
+        Requested => (Running(now.clone(), Duration::from_secs(0)), None),
+
+        // abort sampling => end interpolating
+        Abort => (Done, None),
+
+        // sampling is running, update duration and check end condition
+        Running(last, duration) => {
+            let current_dur = duration + (*now - last);
+            let last_frame = sampler.input.last().cloned().unwrap_or(0.);
+            // duration is past last frame of sampling
+            if last_frame != 0. && duration_to_secs(current_dur) > last_frame {
+                // Check end conditions
+                match control.end {
+                    // Do loop control
+                    EndControl::Loop(times) => {
+                        // Reduce duration down to remaining after last frame
+                        let mut dur_s = duration_to_secs(current_dur);
+                        while last_frame != 0. && dur_s > last_frame {
+                            dur_s -= last_frame;
+                        }
+                        let s = dur_s.floor();
+                        let nanos = dur_s - s;
+                        let next_dur = Duration::new(dur_s as u64, (nanos * 1_000_000_000.) as u32);
+                        match times {
+                            // finite amount of looping, if 0 go do done,
+                            // else reduce iteration count and keep sampling
+                            Some(i) => if i <= 1 {
```

Comment:
Use guards to avoid too many nesting levels

Reply:
Guards?

Reply:
https://rustbyexample.com/flow_control/match/guard.html